packages/blockchain-api/src/currencyConversion/GoldExchangeRateAPI.ts

```diff
@@ -0,0 +1,75 @@
+import { DataSource, DataSourceConfig } from 'apollo-datasource'
+import BigNumber from 'bignumber.js'
+import { database } from '../firebase'
+import { CurrencyConversionArgs } from '../schema'
+import { CUSD } from './consts'
+
+// Firebase stored exchange rate
+interface ExchangeRateObject {
+  exchangeRate: string
+  timestamp: number // timestamp in milliseconds
+}
+
+// Binary search in sorted array
+function findClosestRate(
+  rates: ExchangeRateObject[],
+  timestamp: number
+): ExchangeRateObject | undefined {
+  let lo = 0
+  let hi = rates.length - 1
+  while (lo <= hi) {
+    const mid = Math.floor((lo + hi) / 2)
+    const rate = rates[mid]
+    if (timestamp < rate.timestamp) {
+      hi = mid - 1
+    } else if (timestamp > rate.timestamp) {
+      lo = mid + 1
+    } else {
+      return rate
+    }
+  }
+
+  // At this point lo = hi + 1
+  const loRate = rates[lo]
+  const hiRate = rates[hi]
+  if (!loRate || !hiRate) {
+    return loRate || hiRate
+  }
+
+  return loRate.timestamp - timestamp < timestamp - hiRate.timestamp ? loRate : hiRate
+}
+
+export default class GoldExchangeRateAPI<TContext = any> extends DataSource {
+  // TODO(jeanregisser): memoize results
```

Comment:
TODOs here and below for a future PR? Or left in for now?

Reply:
I left them for a future PR.